src/store/modules/accounts/hdWallet.js

```diff
@@ -29,33 +28,55 @@ export default {
     nextAccountIdx: 1,
   },
   actions: {
-    async isAccountUsed(context, address) {
-      const { getAeSdk } = useAeSdk({ store: context });
-      const aeSdk = await getAeSdk();
-      return aeSdk.api.getAccountByPubkey(address).then(() => true, () => false);
-    },
     async discover({ state, rootGetters, dispatch }) {
       let lastNotEmptyIdx = 0;
       let account;
-      // eslint-disable-next-line no-plusplus
-      for (let nextIdx = state.nextAccountIdx; nextIdx <= 5; nextIdx++) {
-        account = getHdWalletAccount(rootGetters.wallet, nextIdx);
+
+      // Recover Aeternity accounts
+      for (let nextIdx = state.nextAccountIdx; nextIdx <= 5; nextIdx += 1) {
+        const aeternityAdapter = ProtocolAdapterFactory.getAdapter(PROTOCOL_AETERNITY);
+
+        account = aeternityAdapter
+          .getHdWalletAccount(rootGetters.wallet(PROTOCOL_AETERNITY), nextIdx);
         // eslint-disable-next-line no-await-in-loop
-        if (await dispatch('isAccountUsed', account.address)) lastNotEmptyIdx = nextIdx;
+        if (await aeternityAdapter.isAccountUsed(account.address)) {
+          lastNotEmptyIdx = nextIdx;
+        }
       }
-      // eslint-disable-next-line no-plusplus
-      for (let i = state.nextAccountIdx; i <= lastNotEmptyIdx; i++) {
-        dispatch('create', true);
+      for (let i = state.nextAccountIdx; i <= lastNotEmptyIdx; i += 1) {
+        dispatch('create', {
+          isRestored: true,
+          protocol: PROTOCOL_AETERNITY,
+        });
+      }
+
+      // Recover Bitcoin accounts
+      lastNotEmptyIdx = 0;
+      for (let nextIdx = state.nextAccountIdx; nextIdx <= state.nextAccountIdx + 1; nextIdx += 1) {
+        const bitcoinAdapter = ProtocolAdapterFactory.getAdapter(PROTOCOL_BITCOIN);
+
+        account = bitcoinAdapter.getHdWalletAccount(rootGetters.wallet(PROTOCOL_BITCOIN), nextIdx);
+        // eslint-disable-next-line no-await-in-loop
+        if (await bitcoinAdapter.isAccountUsed(account.address)) {
+          lastNotEmptyIdx = nextIdx;
+        }
+      }
+      for (let i = state.nextAccountIdx; i <= lastNotEmptyIdx; i += 1) {
+        dispatch('create', {
+          isRestored: true,
+          protocol: PROTOCOL_BITCOIN,
+        });
       }
     },
```

Comment:
You are repeating those loops for each protocol - maybe they could be extracted to protocol specific functions and used here or to some base functions as this code differs only with the protocol argument
```
for (let i = state.nextAccountIdx; i <= lastNotEmptyIdx; i += 1) {
        dispatch('create', {
          isRestored: true,
          protocol: PROTOCOL_BITCOIN,
        });

Reply:
This has been improved drastically, Please have a look here https://github.com/aeternity/superhero-wallet/pull/2219/files#diff-37085670e4cc4d0853535bb598e3c36b31ffb49af1badbebe2018c1591d51b9c

Resolving this, if you got suggestions for the new approach then please open a new comment.

@pstachel